### ISSUE TITLE
Add Lua 5.2 __gc and __len table metamethods

### ISSUE THREAD

**Issue by Kefta** [Facepunch/garrysmod-requests#585]
>Why doesn't the __gc and __len metamethods work on tables?
>
Userdata objects frequently require some explicit destructor to run when the object is about to be deleted, and Lua provides the __gc metamethod for that purpose. However, this is not allowed on tables for efficiency.
Normally, there is no need to set a destructor on a table, since the table will be deleted automatically, and any references the table contains will then be garbage collected normally. A possible workaround is to create a userdata; make the table the userdata's environment table, and place a reference to the userdata in the table. (Make sure that this is the only reference to the userdata.) When the table becomes collectable, the userdata's __gc metamethod will be run; Lua will not actually destroy the table before that happens because the table is referenced by the userdata.
>
__len on tables is scheduled to be supported in 5.2.

Source: http://lua-users.org/wiki/LuaFaq

**Comment by willox** [comment]
```c
// luaconf.h
#define LUAJIT_ENABLE_LUA52COMPAT
```

* __goto is a keyword and not a valid variable name anymore.__
* break can be placed anywhere. Empty statements (;;) are allowed.
* __lt, __le are invoked for mixed types.
* __len for tables. rawlen() library function.
* pairs() and ipairs() check for __pairs and __ipairs.
* coroutine.running() returns two results.
* table.pack() and table.unpack() (same as unpack()).
* io.write() and file:write() return file handle instead of true.
* os.execute() and pipe:close() return detailed exit status.
* debug.setmetatable() returns object.
* debug.getuservalue() and debug.setuservalue().
* __Remove math.mod(), string.gfind().__

Some features offered by Lua 5.2 are great, some not so great.

If somebody wanted to enable only the __len, __pairs, __ipairs, coroutine.running, and debug.getuservalue/debug.setuservalue changes, that'd be swell.

**Comment by Bo98** [comment]
__ipairs is deprecated in Lua 5.3 so it might be a good idea to avoid that if possible.

Though I might be a bit optimistic in GMod ever reaching Lua 5.3.

**Comment by Kefta** [comment]
@Bo98 Why is __ipairs deprecated out of curiosity?

**Comment by neico** [comment]
Because the way ipairs and pairs are handled is optimized in 5.3 and therefore only needs __pairs to control both, for the specifics you might want to search the lua mailing list...

**Comment by Kefta** [comment]
@neico Thank you

**Comment by neico** [comment]
@Kefta I actually need to correct that statement a bit (should try to cover what I have in mind before writing), ipairs was made aware of metamethods which now looks at __index and replaces what __ipairs did (which makes sense as ipairs is just like a array lookup, while pairs can be used in many ways, which justifies it having a dedicated metamethod): https://github.com/lua/lua/blob/master/src/lbaselib.c#L260

relevant versions: lua 5.3.0-alpha & 5.3.0-beta & 5.3.2-rc1 (feel free to search the version commits: https://github.com/lua/lua/commits/master)

Anyway, to actually contribute something to the actual issue:

https://www.lua.org/manual/5.3/manual.html#3.4.7
>  Unless a __len metamethod is given, the length of a table t is only defined if the table is a sequence, that is, the set of its positive numeric keys is equal to {1..n} for some non-negative integer n. In that case, n is its length. Note that a table like
> 
>      {10, 20, nil, 40}
> 
> is not a sequence, because it has the key 4 but does not have the key 3. (So, there is no n such that the set {1..n} is equal to the set of positive numeric keys of that table.) Note, however, that non-numeric keys do not interfere with whether a table is a sequence. 

and

https://www.lua.org/manual/5.3/manual.html#6.6
> Remember that, whenever an operation needs the length of a table, the table must be a proper sequence or have a __len metamethod (see §3.4.7). All functions ignore non-numeric keys in the tables given as arguments. 

meaning, there are cases where you need to manually calculate the table size for extraordinary usage, a good example would be a ENUM object containing values from -1 to 20 where some values are missing or commented out (think of the <ENUM>_COUNT enum that most enums define at the end of each enum block)

**Comment by Kefta** [comment]
Could really use __gc on tables from 5.2 too.

**Comment by Cheatoid** [comment]
Can we have an update on this?

I believe many would appreciate the `__len` and `rawlen` addition, practically, it is the only missing piece really.
I will try to make a recap of the current state.

----

Looking at GLua (lua_shared), it has been updated to `LuaJIT 2.1.0-beta3`,
with `LUAJIT_ENABLE_LUA52COMPAT` (a.k.a. `LJ_52`) still **not enabled**,
but it seems somethings have been manually "enabled" since 2015,
while somethings are still not "enabled"...

Anyways,
> `goto` is a keyword and not a valid variable name anymore.

It has been a keyword in GLua for some time now... Yep, `goto` (and `::label::`) is a thing as of *today* in GMod.
Moving on,

> io.write() and file:write() return file handle instead of true.
> os.execute() and pipe:close() return detailed exit status.

This is completely irrelevant as GLua doesn't expose such functions.
Moving on,

> `debug.getuservalue` / `debug.setuservalue`

GLua does *not* have these as of *today* yet.

> Remove math.mod(), string.gfind()

These functions shouldn't be used anymore (any decent addon is already updated for this change). From what I can tell, these functions are already being deprecated from GLua (see docs: [math.mod](https://wiki.facepunch.com/gmod/math.mod) & [string.gfind](https://wiki.facepunch.com/gmod/string.gfind)), because they are also already gone from LuaJIT 2.

Last but not least, in my opinion `__pairs`(/`__ipairs`) is not a necessary change (or "a hot" feature), as their behavior can already be implemented by hand if you have a need for it, easily done by detouring (`next`/)`pairs`/`ipairs` global functions, followed by calling appropriate function (per-detour) if such a particular field exists on the metatable for the given table..
A similar (but slightly more complicated) "fix-by-hand" can be done for `__gc` too [...]
But, perhaps because of performance penalty it is better let done in C (as mentioned by neico), although it would also make such hacks go away, making the code cleaner.

----

***tldr***, at this point, we are just missing: `__len` (+ `rawlen`) & `__gc` functionality.
Please add these 😄

**Comment by GitSparTV** [comment]
This request was partially accepted, for example we can have both working goto and goto in variable names.

You can use `__len` and `__gc` using `newproxy`

**Comment by Kefta** [comment]
goto NOT being allowed as a variable name is what the compat enables, and __len isn't called for userdata or tables right now. Nothing has changed since I created this.

**Comment by goodusername123** [comment]
Here is a patchset for LuaJIT [2.0.4, 2.0.5, 2.1.0beta3, and 2.1 rolling release] to unconditionally enable `__pairs`, `__ipairs`, `rawlen()`, and tables honoring `__len` (or just `rawlen()` and table `__len`) without the backwards incompatible aspects/changes brought on by enabling `DLUAJIT_ENABLE_LUA52COMPAT` (some of which [aren't even properly documented](https://github.com/LuaJIT/LuaJIT/commit/0d7094f36e62854d882c528efb634210b6c9455d)!): [LuaJIT unconditionally enable __pairs, __ipairs, rawlen, and table __len.rar.zip](https://github.com/user-attachments/files/17928459/LuaJIT.unconditionally.enable.__pairs.__ipairs.rawlen.and.table.__len.rar.zip) (sorry for the nested archive it's to get around GitHub's upload restrictions)


I can't really comment much on the usefulness of `__pairs` and `__ipairs` myself but I will say that future versions of Lua deciding to deprecate (Lua 5.3) and then remove (Lua 5.4) specifically `__ipairs` shouldn't really be much of a concern considering that It's relatively easy to restore in newer versions of PUC Lua (such as 5.4) if needed and both GMod (pre-existing scripts) and LuaJIT (the creator/author hates newer Lua) are pretty much stuck at a base Lua 5.1 language level for compat reasons anyways, so worrying too much about things removed or changed in future versions is kinda silly considering Lua as a language is a moving target and user made GMod scripts already make use of Lua things that where changed in newer versions. If it does end up being a problem for whatever reason it should be possible to just enable `__pairs` without enabling `__ipairs` in LuaJIT though I didn't feel it was worth the effort making and testing a variation just for that so I didn't. One of the only real reasons I made a variation of the patches for just enabling rawlen() and table __len is just because it was easy due to the order and separation in which I made the patches to begin with and for personal'ish use.
